File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

enum gh{
    MIN,MAX,SUM,IS_BST
};
class Solution {
public:
    int ans;
    // vector order -> min , max , sum, isBst
    void lvr(TreeNode* root, map<TreeNode*,vector<int>> &m){
        if(!root){
            m[root] = {INT_MAX,INT_MIN,0,true};
            return;
        }
        lvr(root->left,m);
        lvr(root->right,m);
        
        // if current subtree a bst
        auto l = root->left;
        auto r = root->right;
        vector<int> temp = {
            min(root->val,min(m[l][MIN],m[r][MIN])),
            max(root->val,max(m[l][MAX],m[r][MAX])),
            root->val+m[l][SUM]+m[r][SUM]
        };
    
        if(
            root->val>m[l][MAX]&&
            root->val<m[r][MIN]&&
            m[l][IS_BST]&&m[r][IS_BST]
        ){
            temp.push_back(true);
        }
        else{
            temp.push_back(false);
        }
        m[root] = temp;
        if (temp[IS_BST]){
            ans = max(ans,temp[SUM]);
        }
    }
    
    int maxSumBST(TreeNode* root) {
        ans = 0;
        map<TreeNode*,vector<int>> m;
        lvr(root,m);
        return ans;
        
    }
};

File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        ans = 0
#         isBST, minVal, maxVal, sum
        def solve(node):
            nonlocal ans
            if not node:
                return (True,inf,-inf,0)
            # if not node.left and not node.right: 
            #     return (True,node.val,node.val,node.val)
            L = solve(node.left)
            R = solve(node.right)
            if L[0] and R[0] and L[2]<node.val<R[1]:
                tempSum = L[3]+R[3]+node.val
                ans = max(ans,tempSum)
                return (True,min(node.val,L[1],R[1]),max(node.val,L[2],R[2]),tempSum)
            return tuple([False])
        
        solve(root)
        return ans